atlascope/core/job_types/nucleus_to_gland_distance.py

```diff
@@ -14,70 +17,108 @@
 
+    # TODO: remove this warning suppression
+    warnings.filterwarnings("ignore")
+
     original_dataset = Dataset.objects.get(id=original_dataset_id)
     job = Job.objects.get(id=job_id)
 
     try:
-        # If a previous job has nucleus detections, use that information
-        nucleus_detections = Job.objects.filter(
-            original_dataset=original_dataset,
-            job_type='nucleus_detection',
+        # We cannot use previous nucleus detection results because we do not
+        # save the nucleus mask to the database
+        # (we need the nucleus mask for the output image)
+        input_image = skimage.io.imread(
+            io.BytesIO(original_dataset.content.read()),
         )
-        if nucleus_detections.count() > 0:
-            nucleus_detections = list(
-                nucleus_detections[0].resulting_datasets.first().detected_nuclei.all()
-            )
-            nucleus_centroids = [
-                (nucleus.centroid.x, nucleus.centroid.y) for nucleus in nucleus_detections
-            ]
-        else:
-            input_image = skimage.io.imread(
-                io.BytesIO(original_dataset.content.read()),
-            )
-            nucleus_detections = detect_nuclei(input_image)
-            nucleus_centroids = [
-                (nucleus["Identifier.CentroidX"], nucleus["Identifier.CentroidY"])
-                for nucleus in nucleus_detections
-            ]
-
-        gland_map = skimage.io.imread(io.BytesIO(open(gland_map_path, 'rb').read()))
-        gland_detections = detect_nuclei(gland_map)
+        nucleus_detections, nucleus_mask = detect_nuclei(input_image)
+        nucleus_centroids = [
+            (nucleus["Identifier.CentroidX"], nucleus["Identifier.CentroidY"])
+            for nucleus in nucleus_detections
+        ]
+
+        gland_map = generate_gland_map(input_image)
+        gland_detections, gland_mask = detect_nuclei(gland_map)
         gland_centroids = [
             (gland["Identifier.CentroidX"], gland["Identifier.CentroidY"])
             for gland in gland_detections
         ]
 
-        distances = [
-            min([math.dist(nucleus_centroid, gland_centroid) for gland_centroid in gland_centroids])
-            for nucleus_centroid in nucleus_centroids
-        ]
+        distances = []
+        mask_shape = (input_image.shape[1], input_image.shape[0])
+        distances_mask = PIL.Image.new("RGB", mask_shape)
+        distances_mask_draw = PIL.ImageDraw.Draw(distances_mask)
+        for nucleus_centroid in nucleus_centroids:
+            minimum = None
+            closest = None
+            for gland_centroid in gland_centroids:
+                distance = math.dist(nucleus_centroid, gland_centroid)
+                if not minimum or distance < minimum:
+                    minimum = distance
+                    closest = gland_centroid
+            distances.append(minimum)
+            distances_mask_draw.line([nucleus_centroid, closest], width=1)
```

Comment:
This is *O(mn)* so if we find it's too slow for our examples, we can apply some acceleration structures here. A simple one is to split the image up into a grid, then only check nuclear-gland distances in the grid region occupied by the nucleus, and the eight neighboring regions. (This is a bit tricky, since if e.g. there are no glands in the nine regions, then you'd have to expand the search until you find one; this complexity is why we should only turn to this approach if we run into performance issues here.)